In [1]:
# Set-up Python libraries - you need to run this but you don't need to change it
import numpy as np
import matplotlib.pyplot as plt
import scipy.stats as stats
import pandas as pd
import seaborn as sns
sns.set_theme(style='white')
import statsmodels.api as sm
import statsmodels.formula.api as smf

# Power

**1. Which effect size requires more participants to detect, d=0.4 or d=0.5?**

0.4, smaller effect sizes need larger n

**2.	For an effect size of d=0.5, which alpha value requires more participants to achieve 80% power, 0.05 or 0.01?**

0.01, smaller alpha requires more participants for the same power

**3.	Testing for an effect size of r=0.37, with a sample size of 50 pairs, what is the power (at alpha=0.05) for 
a.	a one-tailed test and 
b.	a two tailed test? 
Briefly explain the difference in power.**


In [2]:
# import required modules
from statsmodels.stats.power import TTestPower

r = 0.37
n = 50
t = (r*(n-2)**0.5)/((1-r**2)**0.5) # note **2 means 'squared', **0.5 means 'square root')
d = t/(n**0.5)
print('d=' + str(d))

analysis = TTestPower()

# solve for power (one tail)
power = analysis.solve_power(effect_size=d, alpha=0.05, nobs=n, power=None, alternative='larger')
print('power for one-tailed= ' + str(power*100) + '%')

# solve for power (two tail)
power = analysis.solve_power(effect_size=d, alpha=0.05, nobs=n, power=None, alternative='two-sided')
print('power for two-tailed= ' + str(power*100) + '%')

d=0.390217536008329
power for one-tailed= 85.90317965658531%
power for two-tailed= 77.18207898886456%


power is higher for a one-tailed test as, given the direction of effect is known, a lower difference of means is needed to achieve statistical significance for a one-tailed test tahn a two-tailed test

**4.	The heights of 10 English adult men and 10 Dutch adult men are measured. The means and standard deviations for the groups are as follows:**
* English: mean = 177cm, sd = 7.7cm
* Dutch: mean = 184cm, sd = 6.5cm

**a. What is the effect size for the difference of means?**


In [3]:

x1 = 177
x2 = 184

s1 = 7.7
s2 = 6.5

n1 = 10
n2 = 10

s=(((n1-1)*(s2**2) + (n2-1)*(s1**2))/(n1+n2-2))**0.5 # **0.5 means 'to the power of a half' ie square root
s

# Cohen's d
d=(x1-x2)/s
print('d = ' + str(abs(d)))

d = 0.982413808873468


**b.	What is the t-value (you can work it out using the equations!)**

$$t = \frac{\bar{x_1}-\bar{x_2}}{s\sqrt{\frac{1}{n_1}+\frac{1}{n_2}}}$$


In [4]:
t=(x1-x2)/(s*((1/n1)+(1/n2))**0.5)
print('t = ' + str(abs(t)))

t = 2.1967440586755607


Note I am reporting the positive t-value but either sign is OK as it is a 2-tailed test

**c.	A two-tailed t-test for difference of means is carried out at the alpha=5% level. The critical t-value is t=2.1 (that is, the test is significant if it is greater than 2.1)**

**i.	Assuming the null hypothesis is true, what is the probability of a false positive?**

5% as the alpha value is 0.05


**ii.	Assuming the population effect size is the same as the effect size in the sample, what is the probability of a false negative?**

Need a power analysis:

In [5]:
# import required modules
from statsmodels.stats.power import TTestIndPower

# create an analysis of the correct form
analysis = TTestIndPower()

# solve for power
power = analysis.solve_power(effect_size=d, alpha=0.05, nobs1=10, ratio=1, power=None, alternative='two-sided')
print('power = ' + str(power))

power = 0.5473366753981148


Probability of a false negative is 1-0.55 = 45%

**5.	It is sometimes said that significant effects in studies with small sample sizes (underpowered studies) are more likely to be false positives than effects in larger studies. Can you explain why? Key terms to include in your answer are ‘false positive’, ‘power’ and ‘file drawer effect’ – if unsure about the last one (which was mentioned only briefly in the lecture), you can try Googling it!**

The file drawer effect is the phenomenon that studies with statistically significant effects are much more likely to be published. These will be a mixture of true positives and false positives.

When sample size is small, the power is low and therefore the probability of false negatives is high. Conversely the probability of true positives is low.

The probability of false positives if the null is true is fixed at 5% (at alpha=0.05%) regardless of sample size

Therefore, amongst the published (statistically significant results), we expect the same number of false posives for small and large studies, but many fewer true positives for small studies than large studies. Therefore the *proportion* of published studies that are true positives will be lower for small studies. 


# Binomial and Normal Distributions

**1.	A factory produced bags of mini eggs which contain 60 chocolate eggs. We can reasonably assume that the distribution of weights of the bags of eggs is Normal. Explain why, with reference to the central limit theorem.**

The weight of a bag of eggs is the sum of the weights of 60 eggs. If the random variation in weight of individual eggs is independent (which seems reasonable), then the variation in the means of smaples of 60 is normal, by the central limit theorem. But the weight of thje bag of 60 eggs is just 60x the mean weight of the eggs in the bag. So this is also normally distributed.

This is an example of a case in which a measured variable (weights of bags of eggs) is normal because the noise in that measurement arises from a large number of independent sources


**2.	The heights of British women are normally distributed with mean 166cm and standard deviation 6cm.  What is the probability that, in a group of 50 women, at least three are over 175cm tall? (You need to use both the binomial and the normal distributions here)**

In [6]:
#Probability that one is over 175cm
p = 1-stats.norm.cdf(175,166,6)

#probability that two or fewer in n=50 are over 175cm
1-stats.binom.cdf(2,50,p)

0.6577768651225785

66%

**3.	The normal distribution is a good approximation to the binomial when np and nq are both greater than 5. Explain why the approximation breaks down when these conditions are not met (ie, when n is low, or when p is close to 0 or 1).**

The normal-like shape emerges from the number of permutations for each value of k, when n is high.

The shape of the binomial distribution depends on the number of permutations giving each value of k, and the probability of hits and misses within each permutation. When the probability of hits and misses is even (p=0.5) the dostribbution is symmmetrical but as p goes away from 0.5 (towards 1 or 0) the distribution becomes skewed.

As n increases, the skew at a given value of p becomes less prominent, therefore a roughly-symmetrical distribution is obtained even for values of p far from 0.5

**4.	How do we define the ‘best fitting normal distribution’ when fitting a normal approximation to the binomial**


$$k \sim \mathcal{N}(np,\sqrt{npq})$$

The normal has two parameters, mean and sd. If these are set to match the mean and sd of the binomial to be matched, we have done all we can to get the 'best fit'

# Probability theory and Bayes

**For a certain rare disease, 0.1% of the population are infected at any given time. A screening test for the disease measures the level of a certain protein in the blood of the patient. In healthy individuals, the protein level is normally distributed with mean 14300 and standard deviation 1540. In infected individuals, the protein level is normally distributed with mean 27000 and standard deviation 3320.**

**It is decided that if the protein level is above a certain cut-off, patients will be called back for a second test, which is conclusive but involves a painful biopsy.**


**a. Sketch the distribution of protein levels in both healthy and infected individuals on separate graphs.**

**b. Say I want to set the cut-off point such that 99% of infected individuals will be called back for further testing. What is the cut-off point?**

In [7]:
# Note the cutoff point will be *below* the mean for infected individuals*

# Solution 1: use ppf (would have needed to look this up in the help of scipy stats)
print(stats.norm.ppf(0.01,27000,3320))

# Solution 2: change the first argument of cdf until you get the right answer
# approximate start point obtained by rule of thumb that 99% of values lie within 3sd of the mean
print(stats.norm.cdf(19300,27000,3320))

19276.52505818441
0.010190007915349361


**c) Given the cut off point defined in part b, what proportion of healthy individuals will be called back for further testing?**


In [8]:
# note it will be healthy people with scores *over* the cutoff
1-stats.norm.cdf(19277, 14300, 1540)

0.0006150264753376211

0.061%

**4.	A given patient is called back for further testing. What is the probability he has the disease?**

In [9]:
pD = 0.001 # prior probability/base rate for disease
pTgD = 0.99 # p(T|D)=0.99; this is the probability of being called back if infected
pTgDc = 0.000615026 #p(T|Dc); this is the probability of being called back if healthy, as calculated above
pT = (pTgD * pD) + (pTgDc * (1-pD))

# Bayes Theorem
pDgT = (pTgD * pD)/pT

pDgT



0.6170488833866578

62%

**5.	If one million patients are screened, how many would we expect to get an incorrect result from the screening test?**

In [10]:
# false negatives = n * proportion infected * p(false negative given infected) - all given in question
fn = 1000000 * 0.001 * 0.01

# false positives = n * proportion infected * p(false positive given not infected) - latter is from q3
fp = 1000000 * (1-0.001) * 0.000615026

print(fn)
print(fp)
print(fn+fp)



10.0
614.410974
624.410974


624 people will get a wrong result, of whom 10 are false negatives and 614 are false positives

**Consider the screening as a hypothesis testing problem in which the null hypothesis is that the patient is not infected:**

**6. Define a type I and type II error in this context**

Type I is false positive - ie telling the person they are infected when they are not (false alarm, may lead to unnecessary treatment)

Type II is false negative - ie telling the person they are not infected when they are (missing the infection, may lead to failure to treat)


**7.	What would happen to the number of type I and type II errors if a lower protein level was used as the cut-off? Is this desirable?**

Since high levels are indicative of disease, lowering the cutoff would reduce false negatives but increase false positives. It would mean fewer cases are missed, but at the expense of many more people getting unnecessary biopsies. Since there are already 58 unnecessary biopsies per missed case, it may be undesirable to tip the balance even further - but it depends on the consequences of missing a case (will the disease still be treatable if detected later on?)

**8. Without calculation, state which would affect the greater number of patients, the increase in type I or type II errors? Give a reason.**

The increased false positives (Type I) will affect more patients as the disease is rare and the cutoff already gives 58 false positives for each false negaitves

**9. What else would you need to know to decide whether to lower the cutoff point?**

the consequences of missing a case (will the disease still be treatable if detected later on?)

**10. Say the prevalence of the disease in the population was 10%. Without calculation: would the proportion of individuals being mis-diagnosed change? Give a reason.**

Yes, the number of false negatives would increase as the proportion of true cases missed would be fixed, but the number would increase in line with the increase in disease prevalence. The number of false positives would decrease for similar reasons, but not necessarily by a matching amount.